# Introduction
This Workshop is created to give an introduction into the building of Convolutional Neural Network with the help of Tensorflow. 
## Goal
The Goal is to provide a step by step solution to build a model, which seperates good an bad parts on a end-of-line model

![Alt Text](pictures/foto_Modell.png)

## steps
The workshop will include multiple steps to create a model
- inspecting the trainingsdata
- load the data and label it
- build a simple model
- improve the model
- test the model

# Imports
To work create a model and work with it a few imports are necessary

In [4]:
import os
#helpful ml modules
import matplotlib.pyplot as plt
import numpy as np
import cv2
#Tensorflow and keras
import tensorflow as tf
from tensorflow import keras
#create sequential models
from tensorflow.keras.models import Sequential
#import the layers
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, InputLayer, Dropout
#datagenerator
from tensorflow.keras.preprocessing.image import ImageDataGenerator
#augmentation
from tensorflow.keras.layers.experimental import preprocessing

### 1) inspect the datasets
There are two trainingsets in this project, one with 4 subfolders and one with 6 subfolders.
- Which assumptions can you make for the model architecture with this information?
- Load the shape of an image using the cv2 module
- which information can you take away from the shape?

In [3]:
#load the image image = cv2...
image = 
#print the image shape
print(image.shape)

(240, 320, 3)


### 2) creating datasets
To use the given datasets to train a network they need to be loaded. You can do this using cv2 or the keras ImageDataGenrator. Create one Trainings and dataset using both solutions. Decide if you want 4 or 6 classes.

#### 2a) create a dataset using cv2 and for loops
- how many lists do you need?
- how can you get to the images?
- What functions do you need from cv2
- how are you going to label your data? (hint: look at the files or folders)
- how will you normalize the data?
- print the number of images and labels

In [8]:
# trainingsset
# Initialize empty lists for data and labels

# Define the path to your training dataset directory
training_directory = 

# Iterate over subdirectories (classes) in the training dataset directory
for class_name in os.listdir(training_directory):
    class_dir = os.path.join(training_directory, class_name)

    # Iterate over image files in the class directory
    for image_name in os.listdir(class_dir):
        image_path = os.path.join(class_dir, image_name)

        # Load image

        # Normalize pixel values

        # Append image to data list


        # Assign numerical label based on filename.  Here is a if statement needed


# Print the loaded data and labels
print("Number of training samples:", len(x_train))
print("Number of training labels:", len(y_train))

Number of training samples: 840
Number of training labels: 840


expected:\
Number of training samples: 840\
Number of training labels: 840

In [10]:
# testset
# Initialize empty lists for data and labels

# Define the path to your training dataset directory
training_directory = 

# Iterate over subdirectories (classes) in the training dataset directory
for class_name in os.listdir(training_directory):
    class_dir = os.path.join(training_directory, class_name)

    # Iterate over image files in the class directory
    for image_name in os.listdir(class_dir):
        image_path = os.path.join(class_dir, image_name)

        # Load image

        # Normalize pixel values

        # Append image to data list

        # Assign numerical label based on filename. Here is a if statement needed


# Print the loaded data and labels
print("Number of test samples:", len(x_test))
print("Number of test labels:", len(y_test))

Number of test samples: 360
Number of test labels: 360


expected:\
Number of training samples: 360\
Number of training labels: 360\

#### 2b) create a dataset using ImageDataGenerator
- What functions do you need from ImageDatagenerator
- how are you going to label your data? (hint: look at the files or folders)
- how will you normalize the data?
- which configurations can be enabled?
- print the class indices

In [12]:
#trainingset
# traindirectory
data_dir_train = ''

# generate Datengeneratoren
datagen = ImageDataGenerator()

# Training Datengenerator
train_generator = datagen.

class_indices = train_generator.class_indices

# Print the class indices
print("Class indices:", class_indices)

Found 840 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


expected:\
Found 840 images belonging to 4 classes.
Class indices: {'blue': 0, 'fail': 1, 'red': 2, 'white': 3}\
OR\
Found 840 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


In [14]:
# Testset
# testdirectory
data_dir_eval = ''

# configure evaluationgenerator
eval_datagen = ImageDataGenerator()

# create a evaluationset
eval_generator = eval_datagen.
)
class_indices = eval_generator.class_indices

# Print the class indices
print("Class indices:", class_indices)

Found 360 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


expected:\
Found 360 images belonging to 4 classes.
Class indices: {'blue': 0, 'fail': 1, 'red': 2, 'white': 3}\
OR\
Found 360 images belonging to 6 classes.
Class indices: {'blue': 0, 'blue_fail': 1, 'red': 2, 'red_fail': 3, 'white': 4, 'white_fail': 5}


### 3) Create a Classification model
Start with a classification model. This should be able to predict between the Classes you chose (4 or 6). Start to utilize the imported layers of Keras.
- What is the Input dimension?
- Where do you configure the Input?
- How is the last Layer configured?
- Which metric and lossfuntion do you choose?
- where is the flatten layer needed?
- start with the sequential API

In [ ]:
#start to create a Basic sequential model
model = Sequential()
#start adding layers here

#compile your model
opt = keras.optimizers.Adam(learning_rate=0.001)
model.compile(optimizer=opt, loss='', metrics=[''])

### 3a) train the model
- option1: use the two lists generated with the cv2 workflow (think about the batch size)
- option2: use the datagen sets.

In [ ]:
model.fit('''put your traingsset here''', epochs=10)

### 3b)evaluate the model
Evaluate your model, again there are two options.
- How is the accuracy?
- How can it be optimized?

In [ ]:
model.evaluate('''put your traingsset here''')

### 4) Choose optimization methods if needed 
Underfitting (your model cant get a good trainings accuracy)?:
- Research dataaugmentation and add it to either your model or in the dataset (ImageDataGenerator needs to be reconfigured)
- Increase the amount of epochs
- Change the Architecture (More Dense Layers, more Conv2d Layers...)

### 
Overfitting (your model has a good training accuracy, bad a bad validation accuracy)
- Research Dropoutlayers and add them to your model
- research Keras Callback and early stopping

### 5) Export the model to .tflite
if your model has achieved a good validation accuracy you can export it to a tflite model. This can then be loaded to the controller of the conveyor belt.

In [ ]:
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Speichere das tflite-Modell
with open('model.tflite', 'wb') as f:
    f.write(tflite_model)

### 6) load the model on the Controller
- open the RoboCoding App
- connect the Controller to the Computer with USB
- enter the IP 192.168.7.2 in the Browser
- Username: ft Password: fischertechnik
- copy your tflite model to custom/
- change line 19 in the process.py file of the RoboCoding project to your model
- load the programm on the Controller
